generate/templates/manual/src/promise_completion.cc

```diff
@@ -0,0 +1,103 @@
+#include "../include/promise_completion.h"
+
+Nan::Persistent<v8::Function> PromiseCompletion::newFn;
+Nan::Persistent<v8::Function> PromiseCompletion::promiseFulfilled;
+Nan::Persistent<v8::Function> PromiseCompletion::promiseRejected;
+
+// initializes the persistent handles for NAN_METHODs
+void PromiseCompletion::InitializeComponent() {
+  v8::Local<v8::FunctionTemplate> newTemplate = Nan::New<v8::FunctionTemplate>(New);
+  newTemplate->InstanceTemplate()->SetInternalFieldCount(1);
+  newFn.Reset(newTemplate->GetFunction());
+
+  promiseFulfilled.Reset(Nan::New<v8::FunctionTemplate>(PromiseFulfilled)->GetFunction());
+  promiseRejected.Reset(Nan::New<v8::FunctionTemplate>(PromiseRejected)->GetFunction());
+}
+
+bool PromiseCompletion::ForwardIfPromise(v8::Local<v8::Value> result, AsyncBaton *baton, Callback callback)
+{
+  Nan::HandleScope scope;
+
+  // check if the result is a promise
+  if (result->IsObject()) {
+    Nan::MaybeLocal<v8::Value> maybeThenProp = Nan::Get(result->ToObject(), Nan::New("then").ToLocalChecked());
+    if (!maybeThenProp.IsEmpty()) {
+      v8::Local<v8::Value> thenProp = maybeThenProp.ToLocalChecked();
+      if(thenProp->IsFunction()) {
+        // we can be reasonably certain that the result is a promise
+
+        // create a new v8 instance of PromiseCompletion
+        v8::Local<v8::Object> object = Nan::NewInstance(Nan::New(newFn)).ToLocalChecked();
+
+        // set up the native PromiseCompletion object
+        PromiseCompletion *promiseCompletion = ObjectWrap::Unwrap<PromiseCompletion>(object);
+        promiseCompletion->Setup(thenProp.As<v8::Function>(), result, baton, callback);
+
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+// creates a new instance of PromiseCompletion, wrapped in a v8 object
+NAN_METHOD(PromiseCompletion::New) {
+  PromiseCompletion *promiseCompletion = new PromiseCompletion();
+  promiseCompletion->Wrap(info.This());
+  info.GetReturnValue().Set(info.This());
+}
+
+// sets up a Promise to forward the promise result via the baton and callback
+void PromiseCompletion::Setup(v8::Local<v8::Function> thenFn, v8::Local<v8::Value> result, AsyncBaton *baton, Callback callback) {
+  this->callback = callback;
+  this->baton = baton;
+
+  v8::Local<v8::Object> promise = result->ToObject();
+
+  v8::Local<v8::Object> thisHandle = handle();
+
+  v8::Local<v8::Value> argv[2] = {
+    Bind(promiseFulfilled, thisHandle),
```

Comment:
`promiseFulfilled` and `promiseRejected` are instance properties of `PromiseCompletion` objects - https://github.com/srajko/nodegit/blob/8dae154aa00b5c23f482b746ba4efcfb524a812d/generate/templates/manual/src/promise_completion.cc#L13-L14 https://github.com/srajko/nodegit/blob/8dae154aa00b5c23f482b746ba4efcfb524a812d/generate/templates/manual/src/promise_completion.cc#L97-L103

We need to bind `thisHandle` to them so that they can access the `PromiseCompletion` instance here - https://github.com/srajko/nodegit/blob/8dae154aa00b5c23f482b746ba4efcfb524a812d/generate/templates/manual/src/promise_completion.cc#L92-L94

There might be a better way to do the binding though - using the JS `Function.bind` method was the best I could come up with, short of leaking memory by creating a function template for each instance of `PromiseCompletion` :-)